### PR TITLE
New Allocation test [4259]

### PR DESCRIPTION
**Pull request by JuanCarlos-eProsima**
Created a new profiling test to analyze the number of allocations made on each different steps of the execution(Discovery, Communication, Undiscovery).

The main goal is to have the CI system to keep a historical record of the amount of (de)allocations happening after the creation of all entities. This would allow triggering an alarm when an increment is detected.

### CODE REVIEW

**Comment by richiware** [comment]
Build status:
* Linux [Build Status](http://jenkins.eprosima.com:8080/job/FastRTPS%20Manual%20Linux/1063/)
* Mac [Build Status](http://jenkins.eprosima.com:8080/job/FastRTPS%20Manual%20Mac/1033/)
* Windows [Build Status](http://jenkins.eprosima.com:8080/job/FastRTPS%20Manual%20Windows/1047/)

File: test/CMakeLists.txt [hunk 1]
```diff
@@ -40,6 +40,8 @@ if(NOT ((MSVC OR MSVC_IDE)) AND PROFILING_TESTS)
    if(CTEST_MEMORYCHECK_COMMAND)
       add_subdirectory(profiling)
    endif()
+
+   add_subdirectory(profiling/allocations)
```
[the remark this answers is not in the document]
Comment (MiguelCompany) [reply test/CMakeLists.txt]: I think we want the allocations test to be always added, but the more general profiling tests to be added depending on CTEST_MEMORYCHECK_COMMAND. I could be wrong, though.

File: test/profiling/allocations/AllocTestPublisher.cpp [hunk 2]
```diff
@@ -78,51 +92,56 @@ void AllocTestPublisher::PubListener::onPublicationMatched(Publisher* /*pub*/,Ma
...
 }
 
-void AllocTestPublisher::run(uint32_t samples)
+void AllocTestPublisher::run(uint32_t samples, bool wait_unmatch)
 {
-    // Restart callgrind graph
-    callgrind_zero_count();
-
     std::cout << "Publisher waiting for subscriber..." << std::endl;
     while (m_listener.n_matched <= 0)
```
[the remark this answers is not in the document]
Comment (MiguelCompany) [reply test/profiling/allocations/AllocTestPublisher.cpp]: Done

File: test/profiling/allocations/AllocTestPublisher.cpp [hunk 3]
```diff
@@ -78,51 +92,56 @@ void AllocTestPublisher::PubListener::onPublicationMatched(Publisher* /*pub*/,Ma
...
     {
         eClock::my_sleep(25);
     }
 
-    // Flush callgrind graph
-    callgrind_dump();
```
[the remark this answers is not in the document]
Comment (MiguelCompany) [reply test/profiling/allocations/AllocTestPublisher.cpp]: Done

File: test/profiling/allocations/AllocTestPublisher.cpp [hunk 4]
```diff
@@ -78,51 +92,56 @@ void AllocTestPublisher::PubListener::onPublicationMatched(Publisher* /*pub*/,Ma
...
         {
-            std::cout << "Message with index: "<< m_data.index()<< " SENT"<<std::endl;
+            std::cout << "Message with index: "<< m_data.index() << " SENT" << std::endl;
         }
         eClock::my_sleep(500);
 
         if (i == 0)
         {
-            // Flush callgrind graph
-            callgrind_dump();
```
[the remark this answers is not in the document]
Comment (MiguelCompany) [reply test/profiling/allocations/AllocTestPublisher.cpp]: Done

File: test/profiling/allocations/AllocTestPublisher.cpp [hunk 5]
```diff
@@ -78,51 +92,56 @@ void AllocTestPublisher::PubListener::onPublicationMatched(Publisher* /*pub*/,Ma
...
 
-            std::cout << "First message has been sent. Press enter to continue sending samples" << std::endl;
-            std::cin.ignore();
+            std::cout << "First message has been sent" << std::endl;
+            eClock::my_sleep(500);
         }
     }
 
-    // Flush callgrind graph
-    callgrind_dump();
```
[the remark this answers is not in the document]
Comment (MiguelCompany) [reply test/profiling/allocations/AllocTestPublisher.cpp]: Done

File: test/profiling/allocations/AllocTestPublisher.cpp [hunk 6]
```diff
@@ -78,51 +92,56 @@ void AllocTestPublisher::PubListener::onPublicationMatched(Publisher* /*pub*/,Ma
...
+        }
+    }
+    else
+    {
+        std::cout << "All messages have been sent. Waiting a bit to let subscriber receive samples." << std::endl;
+        eClock::my_sleep(500);
+    }
 
-    // Flush callgrind graph
-    callgrind_dump();
```
[the remark this answers is not in the document]
Comment (MiguelCompany) [reply test/profiling/allocations/AllocTestPublisher.cpp]: Done

File: test/profiling/allocations/AllocTestSubscriber.cpp [hunk 7]
```diff
@@ -89,46 +101,51 @@ void AllocTestSubscriber::SubListener::onNewDataMessage(Subscriber* sub)
...
 {
-  run(60);
+  run(60, wait_unmatch);
 }
 
-void AllocTestSubscriber::run(uint32_t number)
+void AllocTestSubscriber::run(uint32_t number, bool wait_unmatch)
 {
-    // Restart callgrind graph
-    callgrind_zero_count();
```
[the remark this answers is not in the document]
Comment (MiguelCompany) [reply test/profiling/allocations/AllocTestSubscriber.cpp]: Done

File: test/profiling/allocations/AllocTestSubscriber.cpp [hunk 8]
```diff
@@ -89,46 +101,51 @@ void AllocTestSubscriber::SubListener::onNewDataMessage(Subscriber* sub)
...
-
     std::cout << "Subscriber waiting for publisher..." << std::endl;
     while (m_listener.n_matched <= 0)
     {
         eClock::my_sleep(25);
     }
 
-    // Flush callgrind graph
-    callgrind_dump();
```
[the remark this answers is not in the document]
Comment (MiguelCompany) [reply test/profiling/allocations/AllocTestSubscriber.cpp]: Done

File: test/profiling/allocations/AllocTestSubscriber.cpp [hunk 9]
```diff
@@ -89,46 +101,51 @@ void AllocTestSubscriber::SubListener::onNewDataMessage(Subscriber* sub)
...
+    eprosima_profiling::discovery_finished();
 
     std::cout << "Subscriber matched. Waiting for first sample..." << std::endl;
     while (this->m_listener.n_samples < 1)
     {
         eClock::my_sleep(25);
     }
 
-    // Flush callgrind graph
-    callgrind_dump();
```
[the remark this answers is not in the document]
Comment (MiguelCompany) [reply test/profiling/allocations/AllocTestSubscriber.cpp]: Done

File: test/profiling/allocations/AllocTestSubscriber.cpp [hunk 10]
```diff
@@ -89,46 +101,51 @@ void AllocTestSubscriber::SubListener::onNewDataMessage(Subscriber* sub)
...
+    eprosima_profiling::first_sample_exchanged();
 
     std::cout << "First sample received. Waiting for rest of samples..." << std::endl;
     while (this->m_listener.n_samples < number)
     {
         eClock::my_sleep(25);
     }
 
-    // Flush callgrind graph
-    callgrind_dump();
```
[the remark this answers is not in the document]
Comment (MiguelCompany) [reply test/profiling/allocations/AllocTestSubscriber.cpp]: Done

File: test/profiling/allocations/AllocTestSubscriber.cpp [hunk 11]
```diff
@@ -89,46 +101,51 @@ void AllocTestSubscriber::SubListener::onNewDataMessage(Subscriber* sub)
...
+    eprosima_profiling::all_samples_exchanged();
 
-    std::cout << "All messages received. Press enter to stop subscriber" << std::endl;
-    std::cin.ignore();
+    if (wait_unmatch)
+    {
+        std::cout << "All messages received. Waiting for publisher to stop." << std::endl;
+        while (m_listener.n_matched > 0)
```
[the remark this answers is not in the document]
Comment (MiguelCompany) [reply test/profiling/allocations/AllocTestSubscriber.cpp]: Done

File: test/profiling/allocations/AllocTestSubscriber.cpp [hunk 12]
```diff
@@ -89,46 +101,51 @@ void AllocTestSubscriber::SubListener::onNewDataMessage(Subscriber* sub)
...
+        }
+    }
+    else
+    {
+        std::cout << "All messages received. Waiting a bit to let publisher receive acks." << std::endl;
+        eClock::my_sleep(500);
+    }
 
-    // Flush callgrind graph
-    callgrind_dump();
```
[the remark this answers is not in the document]
Comment (MiguelCompany) [reply test/profiling/allocations/AllocTestSubscriber.cpp]: Done

**Comment by richiware** [comment]
Build status:
* Linux [Build Status](http://jenkins.eprosima.com:8080/job/FastRTPS%20Manual%20Linux/1064/)
* Mac [Build Status](http://jenkins.eprosima.com:8080/job/FastRTPS%20Manual%20Mac/1034/)
* Windows [Build Status](http://jenkins.eprosima.com:8080/job/FastRTPS%20Manual%20Windows/1048/)

**Comment by richiware** [comment]
Build status:
* Linux [Build Status](http://jenkins.eprosima.com:8080/job/FastRTPS%20Manual%20Linux/1065/)
* Mac [Build Status](http://jenkins.eprosima.com:8080/job/FastRTPS%20Manual%20Mac/1035/)
* Windows [Build Status](http://jenkins.eprosima.com:8080/job/FastRTPS%20Manual%20Windows/1049/)

**Comment by MiguelCompany** [comment]
CC @dejanpan

File: test/CMakeLists.txt (hunk 1 above)
[the remark this answers is not in the document]
Comment (JuanCarlos-eProsima) [reply test/CMakeLists.txt]: We have moved the instruction from this CMakeList file to the one inside the profiling folder.

**Comment by richiware** [comment]
Build status:
* Linux [Build Status](http://jenkins.eprosima.com:8080/job/FastRTPS%20Manual%20Linux/1073/)
* Mac [Build Status](http://jenkins.eprosima.com:8080/job/FastRTPS%20Manual%20Mac/1043/)
* Windows [Build Status](http://jenkins.eprosima.com:8080/job/FastRTPS%20Manual%20Windows/1057/)

**Comment by richiware** [comment]
Build status:
* Linux [Build Status](http://jenkins.eprosima.com:8080/job/FastRTPS%20Manual%20Linux/1097/)
* Mac [Build Status](http://jenkins.eprosima.com:8080/job/FastRTPS%20Manual%20Mac/1067/)
* Windows [Build Status](http://jenkins.eprosima.com:8080/job/FastRTPS%20Manual%20Windows/1081/)

**Pull request merged**